### ISSUE TITLE
RFC: Introduce "block" syntax for all declarations and statement blocks

### ISSUE THREAD

**Issue by bitwalker** [0xPolygonMiden/air-script#331]
This is a proposal to modify the grammar of AirScript to use `{`/`}` to delimit regions and blocks in the language, referred to here as "block syntax". This was first brought up when we discussed solutions to grammar ambiguities with match constraint syntax, which we've since addressed without the changes proposed here (see #328), but had broad support from the team in general. There are additional benefits beyond solving that specific issue though, that I think merit consideration:

* Introducing block syntax would make the language more regular (easier to parse, both mechanically and visually) and would make future extensions to the syntax easier and less likely to run into grammar ambiguities the way match constraint syntax did.
* It will feel more familiar to developers coming from other languages. While AirScript takes some inspiration from Python, AirScript programs don't have much in common with Python at all, instead it is more of a declarative configuration language, e.g. Terraform/HCL, Jsonnet/Dhall/CUE, or even a logic language like Prolog. Most of these languages use block syntax, or some other explicit form of delimiter (e.g. Prolog uses `.` to terminate a declaration), in fact this is true of the vast majority of programming languages. I would argue not much is gained by trying to appear more like Python, but that is obviously subjective.
* The AirScript parser does not track indentation, so it is not actually whitespace-sensitive, which makes the choice of a whitespace-sensitive grammar problematic. It is only the fact that the language has so few features that has prevented this from being a bigger issue. It will become increasingly difficult to add features in the future.

## Example

Take this example AirScript code (I didn't put any effort into making this meaningful/correct, it's just demonstrating syntax):

```airscript
def root

const K = 7

trace_columns:
    main: [clk, a, b[4], c]

public_inputs:
    inputs: [0]

integrity_constraints:
    enf test_constraint(b[1..4])
    enf match:
        case b[0] & b[1]: clk' = clk
        case !b[0] & !b[1]: clk' = 1

boundary_constraints:
    enf clk.first = 0

ev test_constraint([x, ys[2]]):
    let y = sum([col^K for col in ys])
    let z = prod([col^K for col in ys])
    enf x = y + z
```

My proposed changes would result in the same code being expressed like so:

```airscript
def root

const K = 7

trace_columns {
    main: [clk, a, b[4], c]
}

public_inputs {
    inputs: [0]
}

integrity_constraints {
    enf test_constraint(b[1..4])
    enf match {
        case b[0] & b[1]: clk' = clk
        case !b[0] & !b[1]: clk' = 1
    }
}

boundary_constraints {
    enf clk.first = 0
}

ev test_constraint([x, ys[2]]) {
    let y = sum([col^K for col in ys])
    let z = prod([col^K for col in ys])
    enf x = y + z
}
```

While this results in one extra character per declaration (the terminating delimiter), it makes it crystal clear where one declaration starts and the next ends. The biggest beneficiary here is the grammar though, blocks make the syntax unambiguous, which mean that things like the arms of `enf match` can themselves be blocks, whereas without block syntax we cannot support that without some hacks in the parser. Of course it may be the case that supporting that specific enhancement isn't needed, but I think it illustrates the point quite well.

The only thing I don't like (and this is true of the current synax as well) is that there isn't a clear distinction between which top-level declarations can contain code, and which are essentially metadata. Unfortunately, I haven't come up with a syntax that I like which seems to be an improvement, so I think unless we get users expressing confusion over it, or someone comes up with something better, we leave it as-is.

/cc @bobbinth

**Comment by bobbinth** [comment]
#146 is probably somewhat related to this.

**Comment by bobbinth** [comment]
Closed by #350.
